tests/analyser/analyser.cpp

```diff
@@ -0,0 +1,362 @@
+
+    analyser->processModel(model);
+
+    EXPECT_EQ_ISSUES_CAUSES(expectedIssues, expectedCauses, analyser);
+
+    EXPECT_EQ(libcellml::AnalyserModel::Type::INVALID, analyser->model()->type());
+}
+
+TEST(Analyser, nonExistingInitialisingVariable)
+{
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("analyser/non_existing_initialising_variable.cellml"));
+
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    const std::vector<std::string> expectedIssues = {
+        "Variable 'x' has an invalid initial value 'k'. Initial values must be a real number string or a variable reference.",
+    };
+    const std::vector<libcellml::Issue::Cause> expectedCauses = {
+        libcellml::Issue::Cause::VARIABLE,
+    };
+
+    auto analyser = libcellml::Analyser::create();
+
+    analyser->processModel(model);
+
+    EXPECT_EQ_ISSUES_CAUSES(expectedIssues, expectedCauses, analyser);
+
+    EXPECT_EQ(libcellml::AnalyserModel::Type::INVALID, analyser->model()->type());
+}
+
+TEST(Analyser, nonInitialisedState)
+{
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("analyser/non_initialised_state.cellml"));
+
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    const std::vector<std::string> expectedIssues = {
+        "Variable 'x' in component 'my_component' of model 'non_initialised_state' is used in an ODE, but it is not initialised.",
+    };
+    const std::vector<libcellml::Issue::Cause> expectedCauses = {
+        libcellml::Issue::Cause::ANALYSER,
+    };
+
+    auto analyser = libcellml::Analyser::create();
+
+    analyser->processModel(model);
+
+    EXPECT_EQ_ISSUES_CAUSES(expectedIssues, expectedCauses, analyser);
+
+    EXPECT_EQ(libcellml::AnalyserModel::Type::UNDERCONSTRAINED, analyser->model()->type());
+}
+
+TEST(Analyser, underconstrained)
+{
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("analyser/underconstrained.cellml"));
+
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    const std::vector<std::string> expectedIssues = {
+        "Variable 'x' in component 'my_component' of model 'my_model' is not computed.",
+    };
+    const std::vector<libcellml::Issue::Cause> expectedCauses = {
+        libcellml::Issue::Cause::ANALYSER,
+    };
+
+    auto analyser = libcellml::Analyser::create();
+
+    analyser->processModel(model);
+
+    EXPECT_EQ_ISSUES_CAUSES(expectedIssues, expectedCauses, analyser);
+
+    EXPECT_EQ(libcellml::AnalyserModel::Type::UNDERCONSTRAINED, analyser->model()->type());
+}
+
+TEST(Analyser, overconstrained)
+{
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("analyser/overconstrained.cellml"));
+
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    const std::vector<std::string> expectedIssues = {
+        "Variable 'x' in component 'my_component' of model 'my_model' is computed more than once.",
```

Comment:
I find this a little unclear. Could it just be called "over-constrained" instead of "computed more than once"?

Reply:
Well... some people who are not super familiar with modelling might wonder what is meant by over-constrained, under-constrained or unsuitably constrained, hence I thought I would be explicit. @hsorby and @nickerso: views?

Reply:
Then "computed more than once" could be better.  For those of us from the discrete maths background it sounds totally fine :D

Reply:
To be honest, I would not see why computing a variable more than once is an issue. That sounds like something that is done all the time, so this error would actually confuse me more than it helps clarify it. And for an integration, I'd very much hope that all my variables are computed more than once.

Under, over, unsuitably constrained are standard terms that you'd expect libcellml users (tool developers) to be familiar with. Tools would obviously, if needed, present errors in a way that makes most sense to their users.

Reply:
> To be honest, I would not see why computing a variable more than once is an issue.

?? What if you have something like:
```
x = a+b
x = c+d
```

What are we supposed to do with that?

> That sounds like something that is done all the time, so this error would actually confuse me more than it helps clarify it. And for an integration, I'd very much hope that all my variables are computed more than once.

Sure, but you know what I mean about computed more than once... or do you really not? :)

> Under, over, unsuitably constrained are standard terms that you'd expect libcellml users (tool developers) to be familiar with. Tools would obviously, if needed, present errors in a way that makes most sense to their users.

I would probably agree when it comes to users of libCellML, but what about modelling environments like OpenCOR. Clearly, I would expect OpenCOR just to propagate that message to the user, and not all of its users are computational biologists. Truth be told, I was confused the first time I saw the term "over-constrained", etc. in PCEnv/OpenCell.

Reply:
integrating through time, you are repeatedly computing variables over and over and over...

> I would probably agree when it comes to users of libCellML, but what about modelling environments like OpenCOR. Clearly, I would expect OpenCOR just to propagate that message to the user, and not all of its users are computational biologists. Truth be told, I was confused the first time I saw the term "over-constrained", etc. in PCEnv/OpenCell.

exactly, and hence the

> Tools would obviously, if needed, present errors in a way that makes most sense to their users.

I don't see why tools would simply pass libCellML error messages through to users if that would not be useful in their context.